sql/hive-thriftserver/src/test/scala/org/apache/spark/sql/hive/thriftserver/SharedThriftServer.scala

```diff
@@ -74,18 +86,31 @@ trait SharedThriftServer extends SharedSparkSession {
     // Set the HIVE_SERVER2_THRIFT_PORT to 0, so it could randomly pick any free port to use.
     // It's much more robust than set a random port generated by ourselves ahead
     sqlContext.setConf(ConfVars.HIVE_SERVER2_THRIFT_PORT.varname, "0")
-    hiveServer2 = HiveThriftServer2.startWithContext(sqlContext)
-    hiveServer2.getServices.asScala.foreach {
-      case t: ThriftCLIService if t.getPortNumber != 0 =>
-        serverPort = t.getPortNumber
-        logInfo(s"Started HiveThriftServer2: port=$serverPort, attempt=$attempt")
-      case _ =>
-    }
+    // Set the HIVE_SERVER2_THRIFT_HTTP_PORT to 0, so it could randomly pick any free port to use.
+    // It's much more robust than set a random port generated by ourselves ahead
+    sqlContext.setConf(ConfVars.HIVE_SERVER2_THRIFT_HTTP_PORT.varname, "0")
+    sqlContext.setConf(ConfVars.HIVE_SERVER2_TRANSPORT_MODE.varname, mode.toString)
+
+    try {
+      hiveServer2 = HiveThriftServer2.startWithContext(sqlContext)
+      hiveServer2.getServices.asScala.foreach {
+        case t: ThriftCLIService if t.getPortNumber != 0 =>
+          serverPort = t.getPortNumber
+          logInfo(s"Started HiveThriftServer2: port=$serverPort, attempt=$attempt")
```

Comment:
so we may not output this log?

Reply:
Before this fix, yes. The port binding is in another background thread.

Reply:
how does this patch fix it? It seems you just added a try-catch?

Reply:
With https://github.com/apache/spark/pull/28751/files#diff-7610697b4f8f1bc4842c77e50807914cR178 and its implementations, the port binding is done in the same thread where we call `getPortNumber` later.

Reply:
https://github.com/apache/spark/pull/28651#discussion_r435932239 there is a discussion with @juliuszsompolski before

Reply:
ah I see!

Reply:
Take `ThriftBinaryCLIService` for an example
Before:
  we do `TThreadPoolServer` initialization and `serve` in the same `run` function of the background thread.  Then if we call getPortNumber right after `startWithContext`, concurrency issue will occur. The `portNum` may not reset yet when we call.

After:
 we do `TThreadPoolServer` initialization in the current thread and do `serve` in the `run` function of the background thread.

Reply:
Ah, I see. Nice catch.